api/v0/ingest/schema/types.go

```diff
@@ -0,0 +1,178 @@
+package schema
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/multiformats/go-multihash"
+)
+
+type (
+	Advertisement struct {
+		PreviousID *ipld.Link
+		Provider   string
+		Addresses  []string
+		Signature  []byte
+		Entries    ipld.Link
+		ContextID  []byte
+		Metadata   []byte
+		IsRm       bool
+	}
+	EntryChunk struct {
+		Entries []multihash.Multihash
+		Next    *ipld.Link
+	}
+
+	// TODO: Propose this to go-ipld-prime as a way to reduce boilerplate code when working with linksystem?
+	//       Loader interface {
+	//       	FromNode(ipld.Node) error
+	//       	LoadFrom(ipld.LinkContext, ipld.LinkSystem, ipld.Link) error
+	//       }
+	//       Storer interface {
+	//       	ToNode() (ipld.Node, error)
+	//       	StoreTo(ipld.LinkContext, ipld.LinkSystem) (ipld.Link, error)
+	//       }
+)
+
+// ToNode converts this advertisement to its representation as an IPLD typed node.
+//
+// See: bindnode.Wrap.
+func (a Advertisement) ToNode() (n ipld.Node, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			switch x := r.(type) {
+			case string:
+				err = errors.New(x)
+			case error:
+				err = x
+			default:
+				err = fmt.Errorf("unknown panic: %v", r)
+			}
+		}
+	}()
+	n = bindnode.Wrap(&a, TypeAdvertisement).Representation()
+	return
+}
+
+// FromNode populates the fields of this advertisement from the given node.
+func (a *Advertisement) FromNode(node ipld.Node) error {
+	if node.Prototype() != AdvertisementPrototype {
```

Comment:
When do you expect this to kick in? I would document it.

If it's kicking in often, then I imagine something's not right - perhaps your link loading stuff is using `basicnode.Prototype.Any` when following links, and then you're doing the extra work of converting the untyped node to Advertisement here. At the very least this needs a TODO to avoid the duplicate work at some point in the future.

Reply:
The linksystem is passed around to other libraries like go-legs which might load things using `Any` and I don't have control over that unfortunately. All code in this repo, however uses appropriate prototype when loading.

Reply:
That's fine, I would just document it then. I think @willscott, you, and I could briefly chat about this at some point, so that we can properly plumb a NodePrototypeChooser through libraries like go-legs.